controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -1220,29 +1224,88 @@ public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenChec
+            connectionFuture.whenComplete((connection, e) -> connectionCompleteCallback(request, resultFuture, connection, e));
+            resultFuture.whenComplete((result, e) -> requestCompleteCallback(connectionFuture, e));
+        } catch (Exception e) {
+            resultFuture.completeExceptionally(e);
+        }
+    }
+
+    /**
+     * Connection completion callback method. This is invoked when the future returned by the connection pool completes.
+     * If it succeeded, we will have a connection object where send the request. 
+     * If it failed, the resultFuture is failed with ConnectionFailedException. 
+     * @param request request to send over to segment store.
+     * @param resultFuture Future to complete in case of a connection failure. This future is completed in the reply 
+     *                     processor in successful case.  
+     * @param connection Connection object received upon successful completion of future from the request for new 
+     *                   connection from the pool. 
+     * @param e Exception, if any, thrown from attempting to get a new connection.  
+     */
+    private  void connectionCompleteCallback(WireCommand request, CompletableFuture<?> resultFuture,
+                                             ConnectionWrapper connection, Throwable e) {
+        if (connection == null || e != null) {
+            ConnectionFailedException cause = e != null ? new ConnectionFailedException(e) : new ConnectionFailedException();
+            resultFuture.completeExceptionally(new WireCommandFailedException(cause,
+                    request.getType(),
+                    WireCommandFailedException.Reason.ConnectionFailed));
+        } else {                
+            connection.sendAsync(request, resultFuture);
+        }
+    }
+
+    /**
+     * Request Complete callback is invoked when the request is complete, either by sending and receiving a response from 
+     * segment store or by way of failure of connection. 
+     * This is responsible for returning the connection back to the connection pool. 
+     * @param connectionFuture conection future that when completed successfully holds the connection object taken from the pool. 
+     * @param e Exception, if any, thrown from the request processing. 
+     */
+    private void requestCompleteCallback(CompletableFuture<ConnectionWrapper> connectionFuture, 
+                                         Throwable e) {
+        // when processing completes, return the connection back to connection manager asynchronously.
+        // Note: If result future is complete, connectionFuture is definitely complete. if connectionFuture had failed,
+        // we would not have received a connection object anyway. 
+        if (e != null) {
+            Throwable unwrap = Exceptions.unwrap(e);
+            if (hasConnectionFailed(unwrap)) {
```

Comment:
See my earlier comment. I think any unknown exception should be treated as a failure so that the next caller will receive a clean connection.

Reply:
opened seperate issue https://github.com/pravega/pravega/issues/3577 where we will tackle it.